### PR TITLE
Add CFF2 support.

### PR DESCRIPTION
**Pull request by readroberts**
This adds the changes to cffLib.py and psCharstrings.py to add support for CFF2, and a test. I have not yet removed the global data references, but will do so eventually.

### CODE REVIEW

File: Lib/fontTools/ttLib/tables/C_F_F__2.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	def compile(self, otFont):
+		f = BytesIO()
+		self.cff.compile(f, otFont)
+		return f.getvalue()
+
+	def haveGlyphNames(self):
+		if hasattr(self.cff[self.cff.fontNames[0]], "ROS"):
+			return False  # CID-keyed font
+		else:
+			return True
```
Comment (mashabow) [inline Lib/fontTools/ttLib/tables/C_F_F__2.py]: In CFF2, `haveGlyphNames()` should always return `False`.

**Comment by anthrotype** [comment]
Thank you Read.
I still believe that having a global `isCFF2` variable modify the behavior of a class is not the best way to go. I think it'd be cleaner if we factor all the common behavior in abstract base classes and make concrete sub-classes that add format-specific behavior.
I was thinking we could maybe have a `cff2` branch here on upstream fonttools/fonttools where we work together on that, and merge to master once ready?

**Comment by justvanrossum** [comment]
That global `isCFF2` variable is indeed not good. I think fonttools is otherwise thread-safe, and we should keep it that way.

**Comment by readroberts** [comment]
Keeping a branch with CFF2 separate until we fix the issues is fine by me. I agree the global variable should go away. Left to myself, though, I would get to it after I get a full workflow going for building CFF2 fonts,  which is another month or two away.

File: Lib/fontTools/ttLib/tables/C_F_F__2.py (hunk 1 above)
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/C_F_F__2.py]: > In CFF2, haveGlyphNames() should always return False.

That's also my understanding of the spec where it says "The CFF2 format does not contain glyph names".

@readroberts any reason why you left it like that?
Also it appears from `git grep "haveGlyphNames"` that this function is only defined here but not used anywhere...

**Comment by behdad** [comment]
We might as well think of the CFF2 support as WIP and merge, and fix as we go...  Branch is nice, but always more work.

**Comment by readroberts** [comment]
I would argue that a  better name for  'haveGlyphNames' would have been "isNotCID-Keyed", and hence 'haveGlyphNames()' should always return True for  CFF2, as CID support is removed in CFF2. For CFF2 tables, glyph names are no longer  stored in the CFF2 table, but but glyph names for the CFF2 glyphs still exist - they may be defined in the 'post' table, or inferred from the 'cmap' table, as is done by the 'ttx' tool. I don't know why this function is in C_F_F_.py - I did not write that code - but without knowing where it is used, I thought it best to keep it in C_F_F__2.py. All that said, I have no reason to be attached to the existence of this function, as it is not called by any of the AFDKO tools. We can always remove it, and see if anyone complains.

**Comment by anthrotype** [comment]
according to git blame, @justvanrossum added that method to the CFF table in 2002:
https://github.com/fonttools/fonttools/commit/a2a75b348da46073cdd1dea50915384eabfff7c9#diff-1da091ac5ef1be92c6660bd402a26729

if it's not used anywhere else, and it does not make sense to always return True for CFF2, I think we could remove it.

File: Lib/fontTools/cffLib.py [hunk 2]
```diff
@@ -1484,37 +2030,17 @@ def getChildren(self, strings):
 		return children
 
 
-class FontDictCompiler(TopDictCompiler):
+class FontDictCompiler(DictCompiler):
```
Comment (anthrotype) [inline Lib/fontTools/cffLib.py]: Here it seems like you reverted the patch #744:

https://github.com/fonttools/fonttools/pull/744/files#diff-5c4ae0351e8d367dbe256f28eba7db01R1487

@justvanrossum added that to fix issue #740.

@readroberts Did you do that intentionally?
Or perhaps you based your CFF2 patch on an outdated master?

Comment (justvanrossum) [inline Lib/fontTools/cffLib.py]: Aargh :(

**Comment by readroberts** [comment]
I based it on an outdated master; this was not intentional. Will fix this when I am back in the office tomorrow.

**Comment by anthrotype** [comment]
> I based it on an outdated master; this was not intentional. Will fix this when I am back in the office tomorrow.

OK, thanks. Can you also move the files "C_F_F__2_test.py" and "testdata/C_F_F__2.{bin,ttx}" to "Tests/ttLib/tables" and "Test/ttLib/tables/data" respectively?
Cf. #811

**Comment by readroberts** [comment]
I will also move the test files.

**Comment by anthrotype** [comment]
I manually merged `readroberts/master` branch into a new `cff2` branch, here on upstream fonttools/fonttools:

https://github.com/fonttools/fonttools/tree/cff2

Let's keep working together on the new branch.

**Pull request closed without merging**